lib/p2p/Pool.ts

```diff
@@ -750,36 +750,33 @@ class Pool extends EventEmitter {
     peer.removeAllListeners();
     peer.active = false;
 
-    // if handshake passed and peer disconnected from us for stalling or without specifying any reason -
-    // reconnect, for that might have been due to a temporary loss in connectivity
-    const unintentionalDisconnect =
+    const shouldReconnect =
       (peer.sentDisconnectionReason === undefined || peer.sentDisconnectionReason === DisconnectionReason.ResponseStalling) &&
-      (peer.recvDisconnectionReason === undefined || peer.recvDisconnectionReason === DisconnectionReason.ResponseStalling);
+      (peer.recvDisconnectionReason === undefined || peer.recvDisconnectionReason === DisconnectionReason.ResponseStalling ||
+       peer.recvDisconnectionReason === DisconnectionReason.AlreadyConnected ||
+       peer.recvDisconnectionReason === DisconnectionReason.Shutdown);
     const addresses = peer.addresses || [];
 
-    let lastAddress;
-    if (peer.inbound) {
-      lastAddress = addresses.length > 0 ? addresses[0] : undefined;
-    } else {
-      lastAddress = peer.address;
-    }
-
-    if (peer.nodePubKey && unintentionalDisconnect && (addresses.length || lastAddress)) {
+    if (!peer.inbound && peer.nodePubKey && shouldReconnect && (addresses.length || peer.address)) {
       this.logger.debug(`attempting to reconnect to a disconnected peer ${peer.nodePubKey}`);
-      const node = { lastAddress, addresses, nodePubKey: peer.nodePubKey };
+      const node = { addresses, lastAddress: peer.address, nodePubKey: peer.nodePubKey };
       await this.tryConnectNode(node, true);
     }
   }
 
-  private closePeers = () => {
-    this.peers.forEach(peer => peer.close(DisconnectionReason.Shutdown));
+  private closePeers = async () => {
+    for (const peer of this.peers.values()) {
+      await peer.close(DisconnectionReason.Shutdown);
+    }
   }
 
   private closePendingConnections = async () => {
     for (const peer of this.pendingOutboundPeers.values()) {
       await peer.close();
     }
-    this.pendingInboundPeers.forEach(peer => peer.close());
+    for (const peer of this.pendingInboundPeers) {
+      await peer.close();
```

Comment:
I'm thinking that a minor improvement for this and the `closePeers` method is to await all these promises in parallel. We should be trying to close all these connections at once I believe, not one at a time. Can be a different PR if you'd prefer but I figured that while we're changing these methods it might make sense.

Reply:
You're right, executing them in parallel makes sense. Changed `closePendingConnections`, and also `closePeers`.